### PR TITLE
Don't mute hydration errors forcing client render

### PR DESCRIPTION
**Pull request by gaearon**
*tldr: `suppressHydrationWarning` shouldn't mute warnings that force a client render.*

Before 18 + `createRoot`, all hydration mismatches except prop mismatches were patched up. Prop mismatches were warned about but not patched up. And `suppressHydrationWarning` on the parent was a universal way to suppress messages — about all kinds of mismatches.

In 18 + `createRoot`, mismatches always revert to a clean render up to the closest `<Suspense>` above except for one case. The exception is `suppressHydrationWarning`: it enables patching for text mismatches.

However, this means that you can't debug other kinds of mismatches inside `suppressHydrationWarning`. Insertions and deletions now revert to client render but we still mute the warning because in 17 it was harmless to mute. We've introduced limited production behavior for the prop, but we haven't adjusted the warning messages themselves to show for the case where mismatches are *not* patched up.

This fixes the issue.

### CODE REVIEW

File: packages/react-dom/src/__tests__/ReactDOMFizzSuppressHydrationWarning-test.js [hunk 1]
```diff
@@ -248,9 +248,10 @@ describe('ReactDOMFizzServerHydrationWarning', () => {
         ]);
       }).toErrorDev(
         [
+          'Expected server HTML to contain a matching <span> in <span>',
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMFizzSuppressHydrationWarning-test.js:251]: These were being covered up, making mismatches with prod behavior hard to debug and fix.

File: packages/react-dom/src/__tests__/ReactDOMFizzSuppressHydrationWarning-test.js [hunk 2]
```diff
@@ -248,9 +248,10 @@ describe('ReactDOMFizzServerHydrationWarning', () => {
...
           'An error occurred during hydration. The server HTML was replaced with client content in <div>.',
```
Comment (sebmarkbage) [inline packages/react-dom/src/__tests__/ReactDOMFizzSuppressHydrationWarning-test.js:252]: You still have these though. Isn't it better to have the context here for the ones we check in prod anyway? Is it worth having both? Is it confusing that there are two with subtly different timing?

Maybe it would make more sense to have the DEV only error if it was part of a larger diff across multiple mismatches but right now these are 1:1.

Comment (sebmarkbage) [reply packages/react-dom/src/__tests__/ReactDOMFizzSuppressHydrationWarning-test.js:252]: In other words, should we just move the insert/deletion errors to the onRecoverableError?

Comment (gaearon) [reply packages/react-dom/src/__tests__/ReactDOMFizzSuppressHydrationWarning-test.js:252]: I have another PR in the works for the diff. It's not coalesced though.

Comment (gaearon) [reply packages/react-dom/src/__tests__/ReactDOMFizzSuppressHydrationWarning-test.js:252]: But in either case I think we should fix up the inconsistency first. I.e. your concerns don't seem related to this PR per se, it just aligns the behavior for suppressHydration ones with the rest.

File: packages/react-reconciler/src/ReactFiberHydrationContext.old.js [hunk 3]
```diff
@@ -201,6 +203,7 @@ function warnNonhydratedInstance(returnFiber: Fiber, fiber: Fiber) {
       return;
     }
 
+    const isConcurrentMode = (returnFiber.mode & ConcurrentMode) !== NoMode;
```
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberHydrationContext.old.js:206]: I tend to prefer to put these after the tag check. Especially since you don't need to read this field in all cases. Hoisting it out to encourage the anti-pattern of reading fields into variables but in a future refactor not all fields will be available on every type of tag.

Comment (gaearon) [reply packages/react-reconciler/src/ReactFiberHydrationContext.old.js:206]: >the "follow the style of the surrounding code"-rule takes over.

👀

https://github.com/facebook/react/blob/5f7f5280836a44d4b81c1cb13f51e7482470cf1b/packages/react-reconciler/src/ReactFiberHydrationContext.old.js#L479

File: packages/react-dom/src/client/ReactDOMHostConfig.js [hunk 4]
```diff
@@ -1004,14 +1005,20 @@ export function didNotHydrateInstance(
...
-    if (instance.nodeType === ELEMENT_NODE) {
-      warnForDeletedHydratableElement(parentInstance, (instance: any));
-    } else if (instance.nodeType === COMMENT_NODE) {
-      // TODO: warnForDeletedHydratableSuspenseBoundary
-    } else {
-      warnForDeletedHydratableText(parentInstance, (instance: any));
+  if (__DEV__) {
+    if (
+      parentProps[SUPPRESS_HYDRATION_WARNING] !== true ||
+      (enableClientRenderFallbackOnHydrationMismatch && isConcurrentMode)
```
Comment (sebmarkbage) [inline packages/react-dom/src/client/ReactDOMHostConfig.js:1013]: This case is much simpler so can go first.

We're basically saying that suppressHydrationWarning no longer works to suppress child mismatches and as soon as we remove legacy, the check will go away.

**Review by sebmarkbage — APPROVED** [review]
